File: python/processor/docspec_processor/reader/decorator_reader.py
"""Read docspec decorators from AST nodes."""
# @docspec:module {
#   id: "docspec-py-decorator-reader",
#   name: "Decorator Reader",
#   description: "Reads @doc_* decorator metadata from Python AST nodes, extracting keyword arguments into structured annotation dictionaries.",
#   since: "3.0.0"
# }
import ast
from typing import Any


# @docspec:boundary "DocSpec decorator metadata extraction from AST"
class DecoratorReader:
    # @docspec:method { since: "3.0.0" }
    # @docspec:deterministic
    # @docspec:intentional "Extract all @doc_* decorator keyword arguments from an AST node"
    def read(self, node: ast.ClassDef | ast.FunctionDef) -> dict[str, Any]:
        meta: dict[str, Any] = {}
        for decorator in node.decorator_list:
            if isinstance(decorator, ast.Call) and isinstance(decorator.func, ast.Name):
                name = decorator.func.id
                if name.startswith("doc_"):
                    kwargs = {}
                    for kw in decorator.keywords:
                        if kw.arg and isinstance(kw.value, ast.Constant):
                            kwargs[kw.arg] = kw.value.value
                    meta[name] = kwargs
        return meta

File: typescript/processor/src/reader/decorator-reader.ts
// @docspec:module {
//   id: "docspec-ts-decorator-reader",
//   name: "Decorator Reader",
//   description: "Reads DocSpec decorators (@DocModule, @DocHidden, @DocAudience) from TypeScript AST nodes and extracts their configuration into a structured DecoratorMeta object for the processor pipeline.",
//   since: "3.0.0"
// }

import ts from "typescript";

export interface DecoratorMeta {
  module?: { name?: string; description?: string };
  extra: Record<string, unknown>;
}

export class DecoratorReader {
  /** @docspec:intentional "Reads DocSpec decorators from an AST node and extracts configuration into a DecoratorMeta object" */
  read(node: ts.Node): DecoratorMeta {
    const result: DecoratorMeta = { extra: {} };
    const decorators = ts.canHaveDecorators(node) ? ts.getDecorators(node) : undefined;
    if (!decorators) return result;

    for (const decorator of decorators) {
      if (!ts.isCallExpression(decorator.expression)) continue;
      const name = decorator.expression.expression.getText();

      if (name === "DocModule" && decorator.expression.arguments.length > 0) {
        // Parse module options from first argument
        result.module = this.parseObjectLiteral(decorator.expression.arguments[0]);
      }
      if (name === "DocHidden") {
        result.extra.hidden = true;
      }
      if (name === "DocAudience" && decorator.expression.arguments.length > 0) {
        result.extra.audience = this.parseStringArg(decorator.expression.arguments[0]);
      }
    }

    return result;
  }

  /** @docspec:deterministic */
  private parseObjectLiteral(node: ts.Expression): Record<string, any> {
    if (!ts.isObjectLiteralExpression(node)) return {};
    const result: Record<string, any> = {};
    for (const prop of node.properties) {
      if (ts.isPropertyAssignment(prop) && ts.isIdentifier(prop.name)) {
        const key = prop.name.text;
        if (ts.isStringLiteral(prop.initializer)) {
          result[key] = prop.initializer.text;
        } else if (prop.initializer.kind === ts.SyntaxKind.TrueKeyword) {
          result[key] = true;
        } else if (prop.initializer.kind === ts.SyntaxKind.FalseKeyword) {
          result[key] = false;
        }
      }
    }
    return result;
  }

  /** @docspec:deterministic */
  private parseStringArg(node: ts.Expression): string | undefined {
    if (ts.isStringLiteral(node)) return node.text;
    return undefined;
  }
}
